pennylane_qulacs/qulacs_device.py

```diff
@@ -39,181 +74,233 @@
-            self._circuit.add_gate(gate.merge([
-                gate.RZ(wires[0], par[0]),
-                gate.RY(wires[0], par[1]),
-                gate.RZ(wires[0], par[2])
-            ]))
-        elif operation in ('CRZ', 'Toffoli', 'CSWAP'):
-            mapped_operation = self._operations_map[operation]
-            if callable(mapped_operation):
-                gate_matrix = mapped_operation(*par)
-            else:
-                gate_matrix = mapped_operation
-
-            dense_gate = gate.DenseMatrix(wires, gate_matrix)
-            self._circuit.add_gate(dense_gate)
-        else:
-            mapped_operation = self._operations_map[operation]
-            self._circuit.add_gate(mapped_operation(*wires, *par))
+            self._state = QuantumState(self.num_wires)
 
-    @property
-    def state(self):
-        return self._state.get_vector()
+        self._circuit = QuantumCircuit(self.num_wires)
 
-    def pre_measure(self):
-        self._circuit.update_quantum_state(self._state)
+    def apply(self, operations, **kwargs):
+        rotations = kwargs.get("rotations", [])
+
+        self.apply_operations(operations)
+
+        # Rotating the state for measurement in the computational basis
+        self.apply_operations(rotations)
+
+    def apply_operations(self, operations):
+        """Apply the circuit operations to the state.
+
+        This method serves as an auxiliary method to :meth:`~.QulacsDevice.apply`.
+
+        Args:
+            operations (List[pennylane.Operation]): operations to be applied
+        """
+
+        for op in operations:
+            wires = op.wires
+            par = op.parameters
+
+            mapped_op = self._operation_map[op.name]
+            if op.name.endswith(".inv") and mapped_op:
+                # if an inverse variant of the operation exists
+                try:
+                    mapped_operation = getattr(gate, mapped_op.get_name() + "dag")
+                except AttributeError:
+                    # if the operation is hard-coded
+                    try:
+                        if callable(mapped_op):
+                            mapped_operation = np.conj(mapped_op(*par)).T
+                        else:
+                            mapped_operation = np.conj(mapped_op).T
+                    except TypeError:
+                        # else, redefine the operation as the inverse matrix
+                        def mapped_operation(*p):
+                            # embed the gate in a unitary matrix with shape (2**wires, 2**wires)
+                            g = mapped_op(*p).get_matrix()
+                            mat = reduce(np.kron, [I]*len(wires)).astype(complex)
+                            mat[-len(g):, -len(g):] = g
+
+                            # mat follows PL convention => reverse wire-order
+                            gate_mat = gate.DenseMatrix(wires[::-1], np.conj(mat).T)
+                            return gate_mat
```

Comment:
Should probably just put all of this in a `_get_inverse` method instead? :thinking: